migrate.py

```diff
@@ -205,50 +200,55 @@ def _must_keep(self, snapshots: List[str], snapshot: str, keep_last_n: Optional[
                 continue
             if self._get_zfs_snapshot_size(dataset_name, snapshot) == 0:
                 print(F"Skipping snapshot {dataset_name}@{snapshot} because of zero size.")
                 continue
             if not self._must_keep(snapshots_with_size, snapshot, keep_last_n, keep_weekly_n, keep_monthly_n):
-                print(F"Skipping snapshot {dataset_name}@{snapshot} because it does not need to be kept according to the keep pattern.")
+                print(F"Skipping snapshot {dataset_name}@{snapshot} because it does not need to be kept according to the policy.")
+                continue
+            if snapshot in snapshots_in_restic:
+                print(F"Skipping snapshot {dataset_name}@{snapshot} because it's already migrated.")
                 continue
             return snapshot
         return None
 
-    def _backup_next_snapshot_from_dataset(self, dataset_name, keep_last_n: Optional[int], keep_weekly_n: Optional[int], keep_monthly_n: Optional[int]) -> bool:
+    def _backup_next_snapshot_from_dataset(self, dataset_name, larger_than: str, keep_last_n: Optional[int], keep_weekly_n: Optional[int], keep_monthly_n: Optional[int]) -> bool:
         restic_repo, _ = self._get_repo_name_and_path(dataset_name)
 
         snapshots = self._get_dataset_snapshots(dataset_name)
         snapshots_in_restic = self._get_snapshots_in_restic(restic_repo)
         if self.dry_run:
             snapshots_in_restic.update(self._dry_run_finished_backups)
-        parent_restic_snapshot = None
-        newest_snapshot_in_restic = "00000000"
-        if len(snapshots_in_restic) > 0:
-            newest_snapshot_in_restic = self._find_newest_snapshot_in_restic(snapshots_in_restic)
-            parent_restic_snapshot = snapshots_in_restic[newest_snapshot_in_restic]
-        snapshot = self._find_next_snapshot(dataset_name, snapshots, newest_snapshot_in_restic, keep_last_n, keep_weekly_n, keep_monthly_n)
+
+        snapshot = self._find_next_snapshot(dataset_name, snapshots, snapshots_in_restic, larger_than, keep_last_n, keep_weekly_n, keep_monthly_n)
         if snapshot is None:
             print(f"No further snapshots need to backuped for {dataset_name}.")
-            return False
+            return None
+
+        parent_restic_snapshot = None
+        ancestors_in_restic = [ancestor for ancestor in snapshots_in_restic if ancestor < snapshot]
+        if len(ancestors_in_restic) > 0:
+            parent_restic_snapshot = snapshots_in_restic[max(ancestors_in_restic)]
         self._backup_single_snapshot(dataset_name, snapshot, parent_restic_snapshot)
-        return True
+        return snapshot
 
-    def backup_next_snapshot_from_dataset(self, dataset_name, keep_last_n: Optional[int], keep_weekly_n: Optional[int], keep_monthly_n: Optional[int]):
+    def backup_next_snapshot_from_dataset(self, dataset_name, larger_than: str, keep_last_n: Optional[int], keep_weekly_n: Optional[int], keep_monthly_n: Optional[int]):
```

Comment:
This method is used by the cli frontend in line 312. 
This change breaks that.
Either we require to pass that new parameter in the cli, or we choose a sensible default here (e.g. the newest snapshot in restic), or we remove this method and the corresponding part of the cli

Reply:
Ah I read the next commit now.
It is removed there again.